cmd/nitro/init.go

```diff
@@ -173,7 +186,179 @@ func validateBlockChain(blockChain *core.BlockChain, expectedChainId *big.Int) e
 	return nil
 }
 
-func openInitializeChainDb(ctx context.Context, stack *node.Node, config *NodeConfig, chainId *big.Int, cacheConfig *core.CacheConfig) (ethdb.Database, *core.BlockChain, error) {
+type importantRoots struct {
+	bc      *core.BlockChain
+	roots   []common.Hash
+	heights []uint64
+}
+
+// The minimum block distance between two important roots
+const minRootDistance = 2000
+
+// Marks a header as important, and records its root and height.
+// If overwrite is true, it'll remove any future roots and replace them with this header.
+// If overwrite is false, it'll ignore this header if it has future roots.
+func (r *importantRoots) addHeader(header *types.Header, overwrite bool) error {
+	targetBlockNum := header.Number.Uint64()
+	for {
+		if header == nil || header.Root == (common.Hash{}) {
+			return fmt.Errorf("missing state of pruning target block %v", targetBlockNum)
+		}
+		if r.bc.HasState(header.Root) {
+			break
+		}
+		num := header.Number.Uint64()
+		if num%3000 == 0 {
+			log.Info("looking for old block with state to keep", "current", num, "target", targetBlockNum)
+		}
+		// An underflow is fine here because it'll just return nil due to not found
+		header = r.bc.GetHeader(header.ParentHash, num-1)
+	}
+	height := header.Number.Uint64()
+	for len(r.heights) > 0 && r.heights[len(r.heights)-1] > height {
+		if overwrite {
+			r.roots = r.roots[:len(r.roots)-1]
+			r.heights = r.heights[:len(r.heights)-1]
+		} else {
+			return nil
+		}
+	}
+	if len(r.heights) > 0 && r.heights[len(r.heights)-1]+minRootDistance > height {
+		return nil
+	}
+	r.roots = append(r.roots, header.Root)
+	r.heights = append(r.heights, height)
+	return nil
+}
+
+var hashListRegex = regexp.MustCompile("^(0x)?[0-9a-fA-F]{64}(,(0x)?[0-9a-fA-F]{64})*$")
+
+// Finds important roots to retain while proving
+func findImportantRoots(ctx context.Context, chainDb ethdb.Database, stack *node.Node, nodeConfig *NodeConfig, cacheConfig *core.CacheConfig, l1Client arbutil.L1Interface, rollupAddrs arbnode.RollupAddresses) ([]common.Hash, error) {
```

Comment:
current geth code hard-codes block 0 as genesis that won't be pruned.
We should either fix it in geth or always add nitro genesis here. I think adding it here would be simpler, but there are advantages either way.

Reply:
Added :)